### PR TITLE
Adding scheduling for Vertex Pipelines

### PR DESCRIPTION
**Pull request by htahir1**
## Describe changes

Basically I used the approach to schedule Vertex job outlined [here](https://cloud.google.com/vertex-ai/docs/pipelines/schedule-cloud-scheduler) to create a custom scheduler for it. A few impovements are to be made:

- [x]  Clean up duplicate authentication code
- [x] Figure out all the various authentication pieces
- [x] Enable `start_time`, `end_time` mechanic
- [x] Write docs

Its done

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: .pyspelling-ignore-words [hunk 1]
```diff
@@ -1558,3 +1558,5 @@ zenmlregistry
 zenmlserver
 zenserver
 Kubeflow's
+Zipfile
+ziph
```
Comment (htahir1) [inline .pyspelling-ignore-words:1562]:
```suggestion
ziph
TimeoutError
```

File: pyproject.toml [hunk 2]
```diff
@@ -234,6 +234,7 @@ module = [
     "dash_cytoscape",
     "dash.dependencies",
     "docker.*",
+    "flask.*",
```
[the remark this answers is not in the document]
Comment (htahir1) [reply pyproject.toml:237]: Hmm ok yes. Where should I add it as an integration dependency one? And does this matter generally?

File: docs/book/component-gallery/orchestrators/gcloud-vertexai.md [hunk 3]
```diff
@@ -43,6 +43,12 @@ your stack.
 as part of your stack.
 * The GCP project ID and location in which you want to run your Vertex 
 AI pipelines.
+* The pipeline runner environment needs permissions to create a job in Vertex Pipelines,
+e.g. the `Vertex AI User` role: https://cloud.google.com/vertex-ai/docs/general/access-control#aiplatform.user
+* To run on a schedule, the runner environment also needs permissions to create a Google Cloud
+Function (e.g. with the [cloudfunctions.serviceAgent Role](https://cloud.google.com/functions/docs/concepts/iam))
+and to create a Google Cloud Scheduler (e.g. with the
+[Cloud Scheduler Job Runner Role](https://cloud.google.com/iam/docs/understanding-roles)).
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply docs/book/component-gallery/orchestrators/gcloud-vertexai.md:51]: https://zenml.atlassian.net/browse/ENG-1541 here

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply pyproject.toml:237]: I guess here: https://github.com/zenml-io/zenml/blob/561689c2b2f927badaa43dfef3adbda5da67a9bd/src/zenml/integrations/gcp/__init__.py#L45

Well in general we are trying to keep our core dependencies slim

File: src/zenml/integrations/gcp/google_cloud_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    Job periodically sends POST request to the specified HTTP URI on a schedule.
+
+    Args:
+        project: GCP project ID.
+        region: GCP region.
+        http_uri: HTTP URI of the cloud function to call.
+        body: The body of values to send to the cloud function in the POST call.
+        schedule: Cron expression of the schedule. Defaults to "* * * * *".
+        time_zone: Time zone of the schedule. Defaults to "America/Los_Angeles".
```
Comment (AlexejPenner) [inline src/zenml/integrations/gcp/google_cloud_scheduler.py:52]:
```suggestion
        time_zone: Time zone of the schedule. Defaults to "Etc/UTC".
```

**Review by AlexejPenner — APPROVED** [review]
Other than the Flask dependency this would be good to go from my side

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (htahir1) [reply pyproject.toml:237]: This is a mypy ignores not a dependency addition

[the remark this answers is not in the document]
Comment (AlexejPenner) [reply pyproject.toml:237]: :man_facepalming: my bad then, ignore my comment then

File: src/zenml/integrations/gcp/google_cloud_function.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        "Please do not exit the program at this point..."
+    )
+
+    start_time = time.time()
+    while state == Function.State.DEPLOYING:
+        response = get_cloud_functions_api(
+            credentials=credentials
+        ).get_function(
+            request=GetFunctionRequest(
+                name=parent + "/functions/" + sanitized_function_name
```
Comment (schustmi) [inline src/zenml/integrations/gcp/google_cloud_function.py:169]: Use variable, this is used multiple times

File: docs/book/component-gallery/orchestrators/gcloud-vertexai.md [hunk 6]
```diff
@@ -68,6 +74,44 @@ You can now run any ZenML pipeline using the Vertex orchestrator:
...
+The Vertex Pipelines orchestrator supports running pipelines on a schedule, using
+logic resembling the [official approach recommended by GCP](https://cloud.google.com/vertex-ai/docs/pipelines/schedule-cloud-scheduler).
+
+ZenML utilizes the [Cloud Scheduler](https://cloud.google.com/scheduler) and
+[Cloud Functions](https://cloud.google.com/functions) services to enable scheduling
+on Vertex Pipelines. The following is the sequence of events that happen when running
+a pipeline on Vertex with a schedule:
+
+* Docker image is created and pushed (see above [containerization](../../advanced-guide/pipelines/containerization.md)).
+* Cloud Function created that creates the Vertex Pipeline job when triggered.
```
Comment (schustmi) [inline docs/book/component-gallery/orchestrators/gcloud-vertexai.md:88]:
```suggestion
* Cloud Function is created that creates the Vertex Pipeline job when triggered.
```

File: docs/book/component-gallery/orchestrators/gcloud-vertexai.md [hunk 7]
```diff
@@ -68,6 +74,44 @@ You can now run any ZenML pipeline using the Vertex orchestrator:
...
+* Cloud Scheduler job is created that triggers the Cloud Function on the defined schedule.
+
+Therefore, to run on a schedule, the runner environment needs permissions to create a Google Cloud
+Function (e.g. with the [cloudfunctions.serviceAgent Role](https://cloud.google.com/functions/docs/concepts/iam))
+and to create a Google Cloud Scheduler (e.g. with the
+[Cloud Scheduler Job Runner Role](https://cloud.google.com/iam/docs/understanding-roles)).
```
Comment (schustmi) [inline docs/book/component-gallery/orchestrators/gcloud-vertexai.md:94]: mention that it needs to be able to write to the artifact store as well

File: docs/book/component-gallery/orchestrators/gcloud-vertexai.md [hunk 8]
```diff
@@ -68,6 +74,44 @@ You can now run any ZenML pipeline using the Vertex orchestrator:
...
+
+Here is how to create a scheduled pipeline in ZenML:
+
+```python
+from zenml.config.schedule import Schedule
+
+# Run a pipeline every 5th minute
+# Cron expression generated on `https://crontab.guru`
```
Comment (schustmi) [inline docs/book/component-gallery/orchestrators/gcloud-vertexai.md:102]: Is this here on purpose for people that don't know cron schedules and aren't able to google it?

**Review by schustmi — CHANGES_REQUESTED** [review]

File: docs/book/component-gallery/orchestrators/gcloud-vertexai.md (hunk 8 above)
Comment (htahir1) [reply docs/book/component-gallery/orchestrators/gcloud-vertexai.md:102]: Yes

Comment (schustmi) [reply docs/book/component-gallery/orchestrators/gcloud-vertexai.md:102]: I would suggest making it explicit and providing a clickable link outside of the code box then

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 9]
```diff
@@ -212,7 +255,28 @@ def prepare_pipeline_deployment(
...
+            ):
+                logger.warning(
+                    "Vertex orchestrator only uses schedules with the "
+                    "`cron_expression` property. All other properties "
+                    "are ignored."
+                )
+            if deployment.schedule.cron_expression is None:
+                raise ValueError(
+                    "Property `cron_expression` must be set when passing "
+                    "schedule to a vertex orchestrator."
```
Comment (schustmi) [inline src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:277]:
```suggestion
                    "schedule to a Vertex orchestrator."
```

**Pull request merged**